### PR TITLE
MANTA-4682 Want way to display job status

### PR DESCRIPTION
**Pull request by rjloura**

### CODE REVIEW

File: src/util.rs [hunk 1]
```diff
@@ -8,13 +8,62 @@
...
+#[table_name = "pg_database"]
+struct PgDatabase {
+    datname: String,
+}
+
+pub fn list_databases() -> Result<Vec<String>, Error> {
+    let list_query = "SELECT datname FROM pg_database";
+    let conn = PgConnection::establish(&DB_URL)?;
+
+    //let ret: Result<Vec<String>, Error> = sql_query(list_query)
```
Comment (rhb2) [inline src/util.rs]: ^^^ This isn't needed, is it?

File: src/config.rs [hunk 2]
```diff
@@ -196,14 +211,35 @@ impl Command {
...
+                        config.clone(),
+                    )?;
                 }
+            } else if let Some(status_matches) =
+                sub_matches.subcommand_matches("status")
+            {
+                let uuid: Uuid =
+                    Uuid::parse_str(status_matches.value_of("JOB_ID").unwrap())
+                        .unwrap_or_else(|e| {
+                            println!(
```
Comment (rjloura) [inline src/config.rs]: Output of the status command

File: src/util.rs (hunk 1 above)
Comment (rjloura) [reply src/util.rs]: Nope that's cuft.  Thanks for catching that.

File: src/jobs/evacuate.rs [hunk 3]
```diff
@@ -133,27 +141,88 @@ pub enum EvacuateObjectError {
...
+    let status_check = format!("'{}'", status_strings.join("', '"));
+    let error_check = format!("'{}'", error_strings.join("', '"));
+    let skipped_check = format!("'{}'", skipped_strings.join("', '"));
+
+    // TODO: check if table exists first and if so issue warning.  We may
+    // need to handle this a bit more gracefully in the future for
+    // restarting jobs...
+    conn.execute(r#"DROP TABLE evacuateobjects"#)
+        .unwrap_or_else(|e| {
+            debug!("Table doesn't exist: {}", e);
```
[the remark this answers is not in the document]
Comment (rhb2) [reply src/jobs/evacuate.rs]: I saw the TODO statement above -- that was actually what prompted me to ask this question.      I would at least recommend elaborating on that comment and briefly discuss why a table could exist and in such a case, what it means for us, like you did in your response to me.

File: src/jobs/evacuate.rs [hunk 4]
```diff
@@ -269,15 +372,20 @@ pub struct EvacuateJob {
...
-        db_url: &str,
+        db_name: &str,
         max_objects: Option<u32>,
     ) -> Self {
-        let conn = SqliteConnection::establish(db_url)
-            .unwrap_or_else(|_| panic!("Error connecting to {}", db_url));
+        let conn = match util::create_and_connect_db(db_name) {
+            Ok(c) => c,
+            Err(e) => {
+                println!("Error creating Evacuate Job: {}", e);
```
[the remark this answers is not in the document]
Comment (rhb2) [reply src/jobs/evacuate.rs]: The rebalancer will eventually be a daemon anyway, so I don't know if using println! here is really a tenable solution in the long run.

**Comment by rhb2** [comment]
> > I see several places that have println!() which I assume are there just for debugging for the time being. Recommend replacing those with the appropriate macro, whether it's error, trace, debug, info, warn, crit.
> 
> These are outputs from the new `status` and `list` subcommands meant to go to the console and not a log file. I've marked them accordingly. If you see others let me know.
> 
> > There is something about the pg utility stuff that doesn't quite sit right with me being in util.rs and I know we've discussed this at one point in time or another, but I do think it would make sense to have it in a pg.rs or a rebalancer_pg.rs or something.
> 
> Agreed. I moved them.
> 
> > I know you're not going to like this, but I do think, especially for future use, it makes sense to pull this administrative kind of functionality out of the remora binary and have it compile in to a separate utility like a rebalancer-adm or something. It's my opinion that maintaining a separate program for this has the potential to keep the risk and test bar requirements lower when extending it.
> 
> I agree, and that has been the plan all along. Eventually we'll have an API that remote users will call into and receive a response. For now everything zone related is essentially done via a `rebalancer-adm` command.
> 
> `rebalancer-adm job create evacuate <opts>`, `rebalancer-adm job status <job id>`, `rebalancer-adm job list`
> 
> Ideally in the future we'll have something like:
> 
> ```
> manta-rebalancer /jobs/create -d <some json>     ## create a job (currently create and run)
> manta-rebalancer /jobs/<job uuid>                ## for job status
> manta-rebalancer /jobs                           ## for job list
> ```
> 
> This is the reason why I put the status changes in their own module (technically submodule of job), with this own isolated tests and functionality. There is some coupling between the evacuate job and the status module right now. I mentioned this in passing before, but I think once we start to bring in multiple job actions (besides evacuate) we will want to probably create a master DB with a table of job uuids and some high level metadata about them. OTTOMH I'm thinking something like this:
> 
> Job UUID	Job Action	Table Name	Status
> 28e2e15e-549c-4ebd-bdf3-c6977bdfc442	Evacuate	evacuateobjects	Running
> Then that Job UUID would reference a database for each Job.
> 
> But until we do all that work, we'll need to have some coupling between the `job::evacuate` module and the `job::status` module for testing.

But does all that mean that this utility can't be a separate binary today from the remora binary?

**Comment by rjloura** [comment]
> But does all that mean that this utility can't be a separate binary today from the remora binary?

I dont think it makes sense to have different binaries for each rebalancer sub-command.  Also, having multiple sub-commands for a given component follows the general convention throughout triton and manta.

**Comment by rhb2** [comment]
> > But does all that mean that this utility can't be a separate binary today from the remora binary?
> 
> I dont think it makes sense to have different binaries for each rebalancer sub-command. Also, having multiple sub-commands for a given component follows the general convention throughout triton and manta.

I'd probably disagree with that.  For the first comment, I'd disagree on the basis that I'm not asking if we should have a separate binary for every subcommand.  I never asked for *that*.

The general convention across Manta/Triton to the best of my knowledge is to implement a separate utility <whatever>-adm that performs the  kinds of things that you're attempting to do with a listing of job status.  The fact that something like this is already running in a separate process is a sign (in my opinion) that it should be a separate tool.  In general, unless the subcommand is to specify the mode of it (i.e. agent or zone), to follow convention, it really probably should be in a separate binary.

I already mentioned how changes to that code path and keep the overall testing and risk bar lower, but it also may prevent an operator from accidentally starting an agent or another zone.

**Comment by rjloura** [comment]
> > > But does all that mean that this utility can't be a separate binary today from the remora binary?
> > 
> > 
> > I dont think it makes sense to have different binaries for each rebalancer sub-command. Also, having multiple sub-commands for a given component follows the general convention throughout triton and manta.
> 
> I'd probably disagree with that. For the first comment, I'd disagree on the basis that I'm not asking if we should have a separate binary for every subcommand. I never asked for _that_.
> 
> The general convention across Manta/Triton to the best of my knowledge is to implement a separate utility -adm that performs the kinds of things that you're attempting to do with a listing of job status. The fact that something like this is already running in a separate process is a sign (in my opinion) that it should be a separate tool. In general, unless the subcommand is to specify the mode of it (i.e. agent or zone), to follow convention, it really probably should be in a separate binary.
> 
> I already mentioned how changes to that code path and keep the overall testing and risk bar lower, but it also may prevent an operator from accidentally starting an agent or another zone.

I do think the agent should be a separate binary.  Initially I thought it would make sense for it to be an option to the `remora` / `rebalancer-adm` command, but that doesn't make sense with where we have gone.  

I think job CRUD (which we currently have create and will have read after this wad) should be subcommands of the `remora` / `rebalancer-adm` `job` command.

I do think when we enable the ability to run the remora zone in a server mode (via an API as mentioned above) that we will might want a different binary for that, but even then I am not really convinced.

**Comment by rjloura** [comment]
We also discussed previously the idea of separate crates (cargo workspaces) for the agent and the zone.  I think that would make sense in the short term, and (I believe) would give separate binaries for the agent and the zone, but would be out of scope for this change.

**Comment by rhb2** [comment]
> > > > But does all that mean that this utility can't be a separate binary today from the remora binary?
> > > 
> > > 
> > > I dont think it makes sense to have different binaries for each rebalancer sub-command. Also, having multiple sub-commands for a given component follows the general convention throughout triton and manta.
> > 
> > 
> > I'd probably disagree with that. For the first comment, I'd disagree on the basis that I'm not asking if we should have a separate binary for every subcommand. I never asked for _that_.
> > The general convention across Manta/Triton to the best of my knowledge is to implement a separate utility -adm that performs the kinds of things that you're attempting to do with a listing of job status. The fact that something like this is already running in a separate process is a sign (in my opinion) that it should be a separate tool. In general, unless the subcommand is to specify the mode of it (i.e. agent or zone), to follow convention, it really probably should be in a separate binary.
> > I already mentioned how changes to that code path and keep the overall testing and risk bar lower, but it also may prevent an operator from accidentally starting an agent or another zone.
> 
> I do think the agent should be a separate binary. Initially I thought it would make sense for it to be an option to the `remora` / `rebalancer-adm` command, but that doesn't make sense with where we have gone.
The agent should definitely not be part of `rebalancer-adm`.  It's a service that should be running on the storage node at all times, ultimately managed by SMF, and the registrar should be keeping ZK up to date on its status.  I'm not opposed to it being a separate binary, or crate, but I'd agree that that's a conversation beyond the scope of this change, especially since it's not CRUD related.

I'm also not suggesting that we try to separate out the agent from the zone today either and while I'm not sure why that part is coming up here, I'm willing to file a ticket for that and separate it out in to a separate binary.  That's a piece of cake if we want to go that route before we ship in Q1.

> 
> I think job CRUD (which we currently have create and will have read after this wad) should be subcommands of the `remora` / `rebalancer-adm` `job` command.
If I'm understanding correctly, it sounds like you do agree that functionality like this should be part of a rebalancer-adm then.  If that's the case, then I believe that rebalancer-adm should be born from this change, not later.

> 
> I do think when we enable the ability to run the remora zone in a server mode (via an API as mentioned above) that we might want a different binary for that, but even then I am not really convinced.
To be honest, I'm not even sure why the zone wasn't implemented as a server all along.  I don't think it would have even taken more time to do it that way anyway, but at any rate, this is where we are for now, so I'm not suggesting we do anything about *that*.  I'm suggesting that we should not add more stuff to the zone that we know we'll have to tease apart from it later when we do finally add a rebalancer-adm.

File: src/jobs/status.rs [hunk 5]
```diff
@@ -12,59 +12,62 @@ use super::evacuate::EvacuateObjectStatus;
...
         Ok(c) => c,
-        Err(e) => {
-            return Err(format!(
-                "Error connecting to database ({}).  Is this a valid Job \
-                 UUID: {}?",
-                e, db_name
-            ));
+        Err(_) => {
+            return Err(StatusError::DBExists);
         }
```
[the remark this answers is not in the document]
Comment (rjloura) [reply src/jobs/status.rs]: diesel will return a `BadConnection` error if the DB doesn't exist.  I added a catchall for other `ConnectionErrors`, and log them.  There isn't much detail in the docs around exactly how each error can occur.

I also removed mentions of databases from the user facing error messages.

File: src/jobs/status.rs [hunk 6]
```diff
@@ -12,59 +12,62 @@ use super::evacuate::EvacuateObjectStatus;
...
     };
 
-    let status_vec: Vec<EvacuateObjectStatus> = evacuateobjects
-        .select(status)
-        .get_results(&conn)
-        .expect("DB select error");
+    let status_vec: Vec<EvacuateObjectStatus> =
+        match evacuateobjects.select(status).get_results(&conn) {
+            Ok(res) => res,
+            Err(_) => {
```
[the remark this answers is not in the document]
Comment (rjloura) [reply src/jobs/status.rs]: Added some logging here.

**Pull request merged**